File: examples/client-gen/tictactoe/types/tile.py
from __future__ import annotations
import typing
from dataclasses import dataclass
from construct import Container
import borsh_construct as borsh


class TileJSON(typing.TypedDict):
    row: int
    column: int


@dataclass
class Tile:
    layout: typing.ClassVar = borsh.CStruct("row" / borsh.U8, "column" / borsh.U8)
    row: int
    column: int

    @classmethod
    def from_decoded(cls, obj: Container) -> "Tile":
        return cls(row=obj.row, column=obj.column)

    def to_encodable(self) -> dict[str, typing.Any]:
        return {"row": self.row, "column": self.column}

    def to_json(self) -> TileJSON:
        return {"row": self.row, "column": self.column}

    @classmethod
    def from_json(cls, obj: TileJSON) -> "Tile":
        return cls(row=obj["row"], column=obj["column"])

File: ts-reference/examples/tic-tac-toe/generated-client/types/Tile.ts
import { PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh"

export interface TileFields {
  row: number
  column: number
}

export interface TileJSON {
  row: number
  column: number
}

export class Tile {
  readonly row: number
  readonly column: number

  constructor(fields: TileFields) {
    this.row = fields.row
    this.column = fields.column
  }

  static layout(property?: string) {
    return borsh.struct([borsh.u8("row"), borsh.u8("column")], property)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  static fromDecoded(obj: any) {
    return new Tile({
      row: obj.row,
      column: obj.column,
    })
  }

  static toEncodable(fields: TileFields) {
    return {
      row: fields.row,
      column: fields.column,
    }
  }

  toJSON(): TileJSON {
    return {
      row: this.row,
      column: this.column,
    }
  }

  static fromJSON(obj: TileJSON): Tile {
    return new Tile({
      row: obj.row,
      column: obj.column,
    })
  }

  toEncodable() {
    return Tile.toEncodable(this)
  }
}
